### ISSUE TITLE
Path Selection Properties vs. Connection Migration and Multipath

### ISSUE THREAD

**Issue by philsbln** [taps-api/drafts#38]
When thinking about connection migration and multipath, I disagree that Path Selection Properties can not be changed later on. The question is whether a change to the ~~Transport Parameters~~ Path Selection Properties should trigger connection migration or whether another mechanism is needed.

**Comment by britram** [comment]
IMO: nope, don't do this.

A key consideration here is the architecture for multipath: is a multipath connection implemented as a single, multiple-path aware Protocol Stack, or can a TAPS system cobble multipath together out of disparate Protocol Stacks? We went around and around on this point in Post Sockets and decided to punt, though MP transport protocols *really* do need to be linked to each other for transmission scheduling to work well, so on further reflection I'm in favor of the former answer.

This indicates to me that any "automagic" connection migration done by the TAPS system in reaction to changing selection properties of any kind is a bad idea, since doing it right would require the TAPS implementation itself to have its own migration logic separate from and in addition to the Protocol Stack's own migration.

tl;dr if the application's requirements change dynamically to the point that the criteria for selecting a protocol stack have changed, that's an application level event and the application should have direct control of it through re-initiation or re-rendezvous.

**Comment by philsbln** [comment]
- I think we all agree that we don't want to change Protocol Selection Properties after establishment
- We might want to change/provide additional Path Selection Parameter to achieve two things
   - Assign contents to paths within a connection
   - Migrate a connection to a different path

Let me refine my initial question: 
- Which one of the two things above should we support
- Should the API look like changing Transport Parameters or be something else

**Comment by britram** [comment]
ok. thanks for the reframing, I get the question now. 

Which do we want to suport?

> Assign contents to paths within a connection

IMO we absolutely want to do this. In an earlier revision of Post Sockets, one of the main reasons to get a Path object was to pass it to ~Carrier~Connection.Send in order to force a ~Message~Content to be sent down a given Path.

(I'll go ahead and register @mwelzl's probable objection here: no existing API to an mutlipath transport protocol allows you to do this, yet, although some of the work I know is being done on MPQUIC variants will. In any case, I think we absoultely want to do this, so we should specify it, and show how it could be done in the implementation draft)

> Migrate a connection to a different path

We should absolutely support migration, but it's not clear to me that the application needs to be involved directly in migration in the common case. i.e., I don't see why the application doesn't just supply some path selection parameters, and when new paths become available, the system chooses to add and/or migrate paths depending on the patjh selection properties it already has. 

When the application really needs to be in charge of a migration, because different paths are bound in some way to the semantics of the application itself, then ISTM that's a new Initiate().

**Comment by mirjak** [comment]
Two few quick comments on these points:

> Assign contents to paths within a connection

I don't know why you would/should do that. If you e.g. use MPTCP but you only want to use one path then use TCP instead. Using multipath protocols only makes sense if you actually want to use multiple paths; this can be simultaneously, as fallback, or in migration situations.

> Migrate a connection to a different path

I think here the migration question is a transport problem, however, you might need application input to figure out which paths to establish at all and how/when to use them (simultaneously, fallback, make-before-break, or I guess make-after-break would be possible as well). However these are protocol parameters which are only applicable to protocols that support multipath.

**Comment by theri** [comment]
> Assign contents to paths within a connection

I think our API design should support this or at least not prevent an implementation from doing this. Right now, usually there are only two paths and multipath protocols aren't yet widely used. But who knows how this might change in the future? I could imagine a scenario where most clients by default use 3 network interfaces in parallel and an application really wants to avoid scheduling huge objects over one of them, but not care much otherwise. I would like this to be possible with our API.

> If you e.g. use MPTCP but you only want to use one path then use TCP instead.

But does the application necessarily tell us through the API whether it wants to use multiple paths, e.g., through the transport parameters? Right now, it doesn't. Should it? Maybe in a way similar to my suggestion below.

> Migrate a connection to a different path

I like Mirja's categorization of how to use multiple paths (simultaneously, fallback, make-before-break, or I guess make-after-break would be possible as well -- and perhaps also "do not use multiple paths"?) and I think the application should be able to specify this as another Transport Parameter. This way, it can influence how migration happens, but does not have to be involved in the actual migration.

**Comment by tfpauly** [comment]
Mirja's suggestion of how to handle multipath policy after connection is exactly right, in my opinion. We've been working with multipath APIs for several years now and have deployed it quite widely for our internal services. There are better are worse ways to doing this, and I know from experience that getting involved in managing the interfaces and interface types as a maintenance API is extremely hard to get right. You may disagree, but I would strongly urge that we leave the path selection properties in Pre-Establishment for v1 (London). We can always add, but we should be careful not to add complexity where it will hurt us, especially when we have experience to tell us that certain parts are hard.

The generic transport parameter we expose for multipath policy is an enum of values: Handover (only switch paths when you're losing the first one), Interactive (usually use the main one, but retransmit or failover to secondary paths when the first isn't responsive enough), and Aggregation (simultaneous use for improving bandwidth). This set of options at a minimum must be expressible in the API. I don't think we need to go beyond this for v1.

**Comment by philsbln** [comment]
I am working on a pull request to address Issue #34 and #38 by re-shaping the Transport and Send Parameters and reorganising the Parameters. 

- I want Application Intents as Send Parameters be part in v1 – they can be used as input for, e.g., an MPTCP segment scheduler
- I don't want to include application based segment scheduling in v1

**Comment by abrunstrom** [comment]
I agree we should leave the path selection properties in Pre-Establishment for v1. Any Send Parameters that are related only to paths selection should be left out I think. We should really try to focus on a minimal set to start with.

**Comment by tfpauly** [comment]
I agree with Anna that Send Parameters for the purposes of path selection are *not* in scope for v1. While we have experience with multipath protocols, having different chunks of data have different path selection properties is not a feature I am aware of being implemented anywhere at this time. Based on the architectural principles we began with, we should not be adding knobs for features that are not used at this time. We can always add this later if it becomes a reality.

**Comment by philsbln** [comment]
We implemented per-chunk path selection in our Socket Intents Prototype (based on connection pools, but should work the same with MPTCP). This was the thing where BSD Sockets got really ugly.

My fear is that focusing too much on a minimal set might lead us to something that will be insufficient later on.

In the harmonized-parameters branch, I incorporated path selection hints into the Send Properties. As long as we keep Send Properties flexible, e.g. use key/value pairs and not a static list of Properties, we can add them now or later, whatever helps pushing the transport interface.
Its not consistent enough for a push request yet, but should be ready for discussion sometimes tomorrow.

**Comment by tfpauly** [comment]
A few points:
- Not having this be part of v1 should be fine. We can always add things.
- I agree that it sounds interesting to play with per-chunk path selection, but the fact that we can implement it as an experiment doesn't justify inclusion. I was referring to experience deploying with real applications at a large scale, which we have with MPTCP, but not with per-chunk routing in MPTCP.
- We SHOULD NOT be recommending that APIs give Key/Value pair APIs for properties. That's a language choice, and is very very ugly in some languages. We should instead specify that the list should be extensible. The only case we need to avoid is a statically sized struct. An object with properties that can be extended easily is sufficient, and better for type safety than key/values.

**Comment by britram** [comment]
> We can always add things.

on that point, I think we-the-editors need to get together fairly early next week and try to get -00 revs of these documents out. In doing so, we'll need to pull out anything in the API without backing in the architecture and implementation drafts, with the intention to keep the discussion going (indeed, we should probably start moving said discussion onto taps@ietf.org) for an eventual -01. 

So, I'm milestoning this discussion post-London.

> I agree that it sounds interesting to play with per-chunk path selection

as do I; indeed, any implementation of this we do will have Paths as a first-class manipulable object accessible via Connection. Path awareness and path control in the current Internet architecture is, however, an area of open research (this is indeed why we'd add it to our implementation, `#include <panrg_advertisement.txt>`), so I'm not sure the concept needs to show up in the standards track version of the API.

Indeed, the initial work we've done on it in SCION suggests that path selection as a send property in the transport-to-application interface for any other than (1) debug tooling in path-aware networks, (2)  specialized applications in which the application layer protocol is actually performing transmission control itself (this is sometimes the case in SCION for purely accidental reasons), or (3) demonstrations of how cool our multipath stuff is. The path properties really are more useful for (future) path-aware transport protocols to make scheduling decisions based on higher-order requirements (transport and send parameters) passed down by the application.

**Comment by philsbln** [comment]
I am fine with moving this into post-london milestone and remove the path properties from the send parameters in the parameter harmonising PR I'll file tomorrow.

I still want to keep a few Application Intents and Protocol Parameters in there – the first should do no harm, the second is deployed today with SCTP.

@britram does intents based per-content path selection performed by an allication-external local-policy fall in any of your three categories? This is what we did.

**Comment by britram** [comment]
> intents based per-content path selection performed by an application-external local-policy

kinda sounds like (2), unless you did it for demo purposes, in which case it's clearly (3). :) Although I could see some network architectures where you might want application-layer semantics to be reflected in which devices certain packets pass on the way out of the network. That's still a very specialized application.

We want to a be in a place where the application doesn't need to muck about with the details of the path in order to get the multipath service it wants; we'd rather have a set of semantics meaningful to the application that allows a path-aware transport protocol to make good choices on the application's behalf. I agree that until we get there, we need to hand this control across the interface, because there are precisely zero good path-aware transport protocols (MPTCP is a start, but its goals are hardcoded). But just because we need that interface to help us to develop those transports doesn't mean that part of it needs to be standardized at this time.

**Comment by gorryfair** [comment]
I have read thius thread and I think this drfat needs to leave the path selection properties in Pre-Establishment state - It's always possible to be more exotic and there are clearly cases where this will in future be useful. However, I am really keen to avoid making this API byzantine in complexity, and this is an area where I fear this is great PhD material and poor standards.

**Comment by philsbln** [comment]
I came to the conclusion that this kind of functionality is best capsuled either in a connection pool (see Issue #266) or in some migrate/re-initiate call for most practical use cases.

Closing this issue in favour of #266 and #270
